drivers: wifi: esp: support hostname configuration

mniestroj opened the pull request:
Configure hostname by issuing AT+CWHOSTNAME="<hostname>" command. Do it
just after setting link address, which is used to generate hostname
postfix when CONFIG_NET_HOSTNAME_UNIQUE=y.

mniestroj committed:
drivers: wifi: esp: support hostname configuration

Configure hostname by issuing AT+CWHOSTNAME="<hostname>" command. Do it
just after setting link address, which is used to generate hostname
postfix when CONFIG_NET_HOSTNAME_UNIQUE=y.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

jukkar commented on drivers/wifi/esp/esp.c:
```diff
@@ -722,6 +722,21 @@ static int esp_mgmt_ap_disable(const struct device *dev)
+	modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+		       NULL, 0, cmd, &data->sem_response,
+		       ESP_CMD_TIMEOUT);
+#endif
```
probably does not make much difference but we could add here
```
#else 
     ARG_UNUSED(data);
#endif 
```

mniestroj replied:
Is this related to MISRA? I tried to grep for `ARG_UNUSED` and didn't found anything in Zephyr's `doc` directory. So I would like to understand some background instead of blindly following this advise.

jukkar replied:
Yes, this was related to some MISRA stuff. We have this macro in many places of the code base.

jukkar commented on drivers/wifi/esp/esp.c:
```diff
@@ -722,6 +722,21 @@ static int esp_mgmt_ap_disable(const struct device *dev)
+static void esp_configure_hostname(struct esp_data *data)
+{
+#if defined(CONFIG_NET_HOSTNAME_ENABLE)
+	char cmd[64];
```
I am not fond of these hard coded values, what about putting here
```
char cmd[sizeof("AT+CWHOSTNAME=''") +
         sizeof(CONFIG_NET_HOSTNAME) +
         IS_ENABLED(CONFIG_NET_HOSTNAME_UNIQUE) ? sizeof("0011223344556677") : 0);
```

mniestroj replied:
Then what about adding this to `net/hostname.h`:
```
#if defined(CONFIG_NET_HOSTNAME_ENABLE)
#if defined(CONFIG_NET_HOSTNAME_UNIQUE)
/* Allocate extra space to append MAC address to hostname */
#define NET_HOSTNAME_EXTRA_SPACE (8 * 2)
#else
#define NET_HOSTNAME_EXTRA_SPACE 0
#endif /* CONFIG_NET_HOSTNAME_UNIQUE */

#define NET_HOSTNAME_MAX_SIZE	(sizeof(CONFIG_NET_HOSTNAME) +	\
				 NET_HOSTNAME_EXTRA_SPACE)
#endif /* CONFIG_NET_HOSTNAME_ENABLE */
```
or alternatively (based on your above comment):
```
#if defined(CONFIG_NET_HOSTNAME_ENABLE)
#define NET_HOSTNAME_MAX_SIZE	(sizeof(CONFIG_NET_HOSTNAME) +	\
				 IS_ENABLED(CONFIG_NET_HOSTNAME_UNIQUE) ? \
				 sizeof("0011223344556677") : 0)
#endif /* CONFIG_NET_HOSTNAME_ENABLE */
```
and updating `hostname.c` with:
```
static char hostname[NET_HOSTNAME_MAX_SIZE];
```
Then in ESP we can use:
```
char cmd[sizeof("AT+CWHOSTNAME=''") + NET_HOSTNAME_MAX_SIZE];
```

jukkar replied:
The 2nd alternative looks better, +1 to that.

mniestroj committed:
net: hostname: define public macro representing maximum hostname length

Add a macro in public header that represents maximum hostname string
length without terminating NULL character. This will allow other
modules, such as offloaded network drivers, to know how much space is
needed to allocate in order to fit whole hostname.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>
